packages/suite/src/utils/suite/oauth.ts

```diff
@@ -38,70 +38,125 @@ const openWindowOnAnotherDomain = (
     clearInterval(interval);
     interval = window.setInterval(() => {
-        if (win == null || win.closed) {
+        if (!win || win.closed) {
             window.clearInterval(interval);
             closeCallback();
         }
     }, 1000);
     return win;
 };
 
-/**
- * Handle extraction of authorization code from Oauth2 protocol
- */
-export const extractCredentialsFromAuthorizationFlow = (url: string) => {
-    const originalParams = urlHashParams(url);
-
-    const dfd: Deferred<Credentials> = createDeferred();
+const handleResponse = (
+    message: { [key: string]: string },
+    originalParams: { [key: string]: string },
+    onSuccess: (result: Credentials) => void,
+    onError: (error: any) => void,
+) => {
+    if (!message.search && !message.hash) return;
 
-    const onMessageWeb = (e: MessageEvent) => {
-        // oauth message is post-messaged from oauth_receiver.html that is hosted on the same origin
-        if (window.location.origin !== e.origin) {
-            return;
-        }
+    let parsedMessage;
 
-        if (!e.data.search && !e.data.hash) return;
+    if (message.search) {
+        parsedMessage = urlSearchParams(message.search);
+    } else if (message.hash) {
+        parsedMessage = urlHashParams(message.hash);
+    }
 
-        let message;
+    if (!parsedMessage) return;
+
+    const { code, access_token, state, error_description, error } = parsedMessage;
+
+    if (error === 'access_denied') {
+        // user clicks cancel in popup interface. This is the same as if user closed the window,
+        onError(new Error('window closed'));
+    } else if (error) {
+        // otherwise just show error. This should not happen often, most of the errors can be caused by improper
+        // implementation only. Possibly "temporarily_unavailable" or "server_error" may appear
+        // see possible errors in oauth protocol described here https://tools.ietf.org/html/rfc6749#section-4.1.2.1
+        onError(new Error(`${error}: ${error_description.replace(/\+/g, ' ')}`));
+    } else if (originalParams.state && state !== originalParams.state) {
+        onError(new Error('state does not match'));
+    } else if (code || access_token) {
+        onSuccess(({ code, access_token } as unknown) as Credentials);
+    } else {
+        onError(new Error('Unexpected response form data provider'));
+    }
+};
 
-        if (e.data.search) {
-            message = urlSearchParams(e.data.search);
-        } else if (e.data.hash) {
-            message = urlHashParams(e.data.hash);
-        }
-        if (!message) return;
+// keep handler function instance in top level scope
+let desktopHandlerInstance: (message: { [key: string]: string }) => void;
+let webHandlerInstance: (e: MessageEvent) => void;
 
-        const { code, access_token, state } = message;
+const getDesktopHandlerInstance = (
+    dfd: Deferred<Credentials>,
+    originalParams: { [key: string]: string },
```

Comment:
let's maybe add type as it is used in multiple places

Reply:
well, why not. It has 13 occurrences in total across the project. What would be the name for this utility type?

Reply:
`KeyValueString`?

Reply:
cache invalidation and naming stuff are the hardest things in software development 😅  anyway I'd go with @keraf suggestion

Reply:
And you would expect me to put it into types/utils/index.ts? I am still struggling to see added value of this. If I see `{[key:string]:string}` I immediately know what is going on, I see this one as a common idiom, nearly the same as `number` or `boolean`. When I see `KeyValueString` I really have no idea what to expect. It does not appear to be a good deal to have less typing 14 (KeyValueString) vs 21 ({[key:string]:string}) in exchange of added mental overhead. I'd support your idea in case it was more complicated type, where we could say that it might be edited in future. Here I don't see any chance how it could ever be edited.